src/EventListener/MergeHttpHeadersListener.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2015 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\CoreBundle\ContaoFrameworkInterface;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+
+/**
+ * Merges HTTP headers sent using PHP's header() method into the Symfony
+ * Response so they are available for other listeners following later in the
+ * application flow and relying on headers to be part of the response object.
+ *
+ * Although this could be useful in many applications we think that this should
+ * not be supported in general and only serves as a layer to support legacy
+ * code and therefore it's restricted to the Contao framework and part of the
+ * Contao Core Bundle.
+ *
+ * @author Yanick Witschi <https://github.com/toflar>
+ */
+class MergeHttpHeadersListener
+{
+    /**
+     * @var array
+     */
+    private $routeNames;
+    /**
+     * @var ContaoFrameworkInterface
+     */
+    private $contaoFramework;
+
+    /**
+     * Constructor.
+     *
+     * @param array $routeNames
+     * @param ContaoFrameworkInterface $contaoFramework
+     */
+    public function __construct(array $routeNames, ContaoFrameworkInterface $contaoFramework)
+    {
+        $this->routeNames = $routeNames;
+        $this->contaoFramework = $contaoFramework;
+    }
+
+    /**
+     * Sets the default locale based on the request or session.
+     *
+     * @param GetResponseEvent $event The event object
+     */
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        if (!$this->contaoFramework->isInitialized()) {
+
+            return;
+        }
+
+        $request = $event->getRequest();
+
+        if (!$request->attributes->has('_route')
+            || !in_array($request->attributes->get('_route'), $this->routeNames)
+        ) {
+
+            return;
+        }
+
+        $event->setResponse($this->mergeHttpHeaders($event->getResponse()));
+    }
+
+    /**
+     * Merge the http headers. If one of the headers is already present in the
+     * response, it will not be merged as the response object has a higher
+     * priority.
+     *
+     * @param Response $response
+     *
+     * @return Response
+     */
+    private function mergeHttpHeaders(Response $response)
+    {
+        if ('cli' === PHP_SAPI && function_exists('xdebug_get_headers')) {
+            $headers = xdebug_get_headers();
+        } else {
+            $headers = headers_list();
+        }
+
+        foreach ($headers as $header) {
+            list($name, $content) = explode(':', $header, 2);
+            $content = trim($content);
+
+            if ($response->headers->has($name)) {
+                continue;
+            }
+
+            $response->headers->set($name, $content);
```

Comment:
Headers may be present more than once, therefore I guess we should skip the check and merge the value by using 
```diff
-            if ($response->headers->has($name)) {
-                continue;
-            }
-
-            $response->headers->set($name, $content);
+            $response->headers->set($name, $content, false);
```